File: week05/19/P2.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* insert_head(ListNode *head, int val) {
    ListNode *new_node = new ListNode(val, head);
    return new_node;
};

void printList(ListNode* head) {
    ListNode* current_node = head;
    while (current_node != nullptr) {
        cout << current_node->val << " -> ";
        current_node = current_node->next;
    }
    cout << "nullptr" << endl;
}

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr) return head;
        
        ListNode *temp = head->next;
        ListNode *prev = head;

        while (temp != nullptr) {
            if (temp->val == prev->val) {
                prev->next = temp->next;
            } else {
                prev = temp;
            }
            temp = temp->next;
        }
        return head;
    }
};

int main() {
    ListNode *head = nullptr;
    head = insert_head(head, 3);
    head = insert_head(head, 3);
    head = insert_head(head, 2);
    head = insert_head(head, 1);
    head = insert_head(head, 1);
    printList(head);  // 1 -> 1 -> 2 -> 3 -> 3 -> nullptr
    
    Solution result;
    head = result.deleteDuplicates(head);
    printList(head);  // 1 -> 2 -> 3 -> nullptr

    return 0;
}

File: week05/19/P2.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = ListNode(data)
        new_node.next = self.head
        self.head = new_node

    def print_list(self):
        curr = self.head
        while (curr is not None):
            print(curr.val, end=" -> ")
            curr = curr.next
        print("nullptr")

    def deleteDuplicates(self):
        if (self.head is None):
            return

        temp = self.head.next
        prev = self.head
            
        while (temp is not None):
            if (temp.val == prev.val):
                prev.next = temp.next
            else:
                prev = temp
            temp = temp.next

if __name__ == "__main__":
    head = Solution()
    head.insert(3)
    head.insert(3)
    head.insert(2)
    head.insert(1)
    head.insert(1)
    head.print_list()  # 1 -> 1 -> 2 -> 3 -> 3 -> nullptr

    head.deleteDuplicates()
    head.print_list()  # 1 -> 2 -> 3 -> nullptr